File: src/updatable_encryption_python/parameters.py
import numpy as np

# requirements
# n ≥ 1
# q ≥ 2
# q = poly(λ) - means q should be big so its hard to break
# m ≥ nk ≥ n
# k = ⌈log2q⌉
# m = ¯m+2nk

n = 16
q = 1048573 #8380417 #2647 #2048 #1979  #2647   #1979
k = int(np.ceil(np.log2(q)))
print("k = ", k)
isexactpower = (q == 2**k)
nk = n*k
m_tilda = 32 #nk #* 2 #10
m = m_tilda + 2*nk
alpha = 1/10
s_hidden = np.zeros(n, dtype=int)  #used for testing
e0_hidden = np.zeros(m_tilda, dtype=int)  #used for testing
e1_hidden = np.zeros(nk, dtype=int)  #used for testing
e2_hidden = np.zeros(nk, dtype=int)  #used for testing

e00 = np.zeros(m_tilda, dtype=int)  #used for testing
e01 = np.zeros(nk, dtype=int)  #used for testing
e02 = np.zeros(nk, dtype=int)  #used for testing

def mod_q_vector(vector):
    for current_element in range(len(vector)):

        vector[current_element] = vector[current_element] % q
        while vector[current_element] > q/2:
            vector[current_element] -= q
        while vector[current_element] < -q/2:
            vector[current_element] += q   
    return vector

def updateError(e0, e1, e2):
    global e0_hidden, e1_hidden, e2_hidden
    e0_hidden = e0
    e1_hidden = e1
    e2_hidden = e2
        

def updateError2(e0, e1, e2, M):
    global e0_hidden, e1_hidden, e2_hidden, e00, e01,e02
    helpe1 = np.concatenate((e0_hidden, e1_hidden, e2_hidden))
    helpe1 = np.dot(helpe1, M % q)
    e00 = mod_q_vector(helpe1[:m_tilda])
    e01 = mod_q_vector(helpe1[m_tilda:m_tilda+nk])
    e02 = mod_q_vector(helpe1[m_tilda+nk:])
    e0_hidden = mod_q_vector(helpe1[:m_tilda] + e0)
    e1_hidden = mod_q_vector(helpe1[m_tilda:m_tilda+nk] + e1)
    e2_hidden = mod_q_vector(helpe1[m_tilda+nk:] + e2)



def getError():
    global e0_hidden, e1_hidden, e2_hidden
    return (e0_hidden, e1_hidden, e2_hidden) 

def getError2():
    global e00, e01,e02
    return (e00, e01, e02) 

def updateS(s):
    global s_hidden
    s_hidden = (s + s_hidden) % q
    print("s_hidden: ", s_hidden)

def getS():
    global s_hidden
    return s_hidden    

# normal distribution with following params
mean = 0
sigma_sk =  0.7 # 1 * np.sqrt(np.log(n))  # R dispersion, included in Rte(<= q/4) affecting correct sampling and decoding
edisp = 0.7 # error dispersion , included in Rte(<= q/4) affecting correct sampling and decoding
tau_sample = 2 / (2+1)  # s / (b+1)  recommended by LWE algorithm , affecting correct sampling and decoding

def generate_G(n, k, q):
    # Construct the vector g^t = [1, 2, 4, ..., 2^(k-1)]
    g_t = np.array([2**i for i in range(k)])  # Shape: (k,)
    # Compute G using the Kronecker product G = I_n ⊗ g^t
    I_n = np.eye(n, dtype=int)  # Identity matrix of size (n, n)
    G = np.kron(I_n, g_t)  # Shape: (n, nk)
    #print("Matrix G:\n", (G))
    #type = <class 'numpy.ndarray'>
    return G % q

G = generate_G(n, k, q)

File: chaincode-go/updatable_encryption/parameters.go
/*
# requirements
# n ≥ 1
# q ≥ 2
# q = poly(λ) - means q should be big so its hard to break
# m ≥ nk ≥ n
# k = ⌈log2q⌉
# m = ¯m+2nk
*/

package updatable_encryption

import (
	"fmt"
	"math"
)

type Constants struct {
	n            int
	q            int
	k            int
	isexactpower bool
	nk           int
	m_tilda      int
	m            int
	alpha        float64
	G            [][]int

	mean       float64
	sigma_sk   float64
	edisp      float64
	tau_sample float64

	//testing
	s_hidden  []int
	e0_hidden []int
	e1_hidden []int
	e2_hidden []int
	e00       []int
	e01       []int
	e02       []int
}

func (obj *Constants) InitConstants() {
	obj.n = 16
	obj.q = 1048573 //8380417
	obj.k = int(math.Ceil(math.Log2(float64(obj.q))))
	obj.isexactpower = (obj.q == int(math.Pow(float64(2), float64(obj.k))))
	obj.nk = obj.n * obj.k
	obj.m_tilda = 32
	obj.m = obj.m_tilda + 2*obj.nk
	obj.alpha = float64(1) / float64(10)
	obj.G = generate_G(obj.n, obj.k, obj.q, obj.nk)

	obj.mean = 0
	obj.sigma_sk = 0.7
	obj.edisp = 0.7
	obj.tau_sample = 2 / (2 + 1)

	obj.s_hidden = make([]int, obj.n)
	obj.e0_hidden = make([]int, obj.m_tilda)
	obj.e1_hidden = make([]int, obj.nk)
	obj.e2_hidden = make([]int, obj.nk)
	obj.e00 = make([]int, obj.m_tilda)
	obj.e01 = make([]int, obj.nk)
	obj.e02 = make([]int, obj.nk)
}

func generate_G(n, k, q, nk int) [][]int {
	gT := make([]int, k)
	for i := 0; i < k; i++ {
		gT[i] = int(math.Pow(2, float64(i)))
	}

	G := make([][]int, n)
	for i := 0; i < n; i++ {
		G[i] = make([]int, nk)
		for j := 0; j < k; j++ {
			G[i][i*k+j] = gT[j] % q
		}
	}
	return G
}

var C Constants

func update_error(e0, e1, e2 []int) error {

	if len(C.e0_hidden) != len(e0) || len(C.e1_hidden) != len(e1) || len(C.e2_hidden) != len(e2) {
		return fmt.Errorf("update_error - missmatching sizes")
	}

	for i := range e0 {
		C.e0_hidden[i] = e0[i]
	}

	for i := range e1 {
		C.e1_hidden[i] = e1[i]
	}

	for i := range e2 {
		C.e2_hidden[i] = e2[i]
	}

	return nil
}

func update_error2(e0, e1, e2 []int, M [][]int) error {
	helpe1 := append(append(C.e0_hidden, C.e1_hidden...), C.e2_hidden...)

	helper2, err2 := dot_product_VM(helpe1, M)

	if err2 != nil {
		return fmt.Errorf("update_error2 - err2 is not nil")
	}

	if len(helper2) != C.m {
		return fmt.Errorf("update_error2 - missmatching sizes")
	}

	//helper2 = modQ(helper2)

	for i := range e0 {
		C.e00[i] = helper2[i]
	}

	for i := range e1 {
		C.e01[i] = helper2[i+C.m_tilda]
	}

	for i := range e2 {
		C.e02[i] = helper2[i+C.m_tilda+C.nk]
	}

	helper3, err3 := add_vectors(helper2[:C.m_tilda], e0)
	helper4, err4 := add_vectors(helper2[C.m_tilda:C.m_tilda+C.nk], e1)
	helper5, err5 := add_vectors(helper2[C.m_tilda+C.nk:], e2)

	if err3 != nil || err4 != nil || err5 != nil {
		return fmt.Errorf("update_error2 - err3 or err4 or err5 is not nil")
	}

	//helper3 = modQ(helper3)
	//helper4 = modQ(helper4)
	//helper5 = modQ(helper5)

	if len(C.e0_hidden) != len(helper3) || len(C.e1_hidden) != len(helper4) || len(C.e2_hidden) != len(helper5) {
		return fmt.Errorf("update_error2 - missmatching sizes 2")
	}

	for i := range C.e0_hidden {
		C.e0_hidden[i] = helper3[i]
	}

	for i := range C.e1_hidden {
		C.e1_hidden[i] = helper4[i]
	}

	for i := range C.e2_hidden {
		C.e2_hidden[i] = helper5[i]
	}

	return nil
}

func get_error() ([]int, []int, []int) {
	return C.e0_hidden, C.e1_hidden, C.e2_hidden
}

func get_error2() ([]int, []int, []int) {
	return C.e00, C.e01, C.e02
}

func updateS(s []int) error {

	if len(s) != len(C.s_hidden) {
		return fmt.Errorf("update_s - missmatching sizes")
	}
	for i := range C.s_hidden {
		C.s_hidden[i] = (C.s_hidden[i] + s[i]) % C.q
	}

	//C.s_hidden = s
	return nil
}

func getS() []int {
	return C.s_hidden
}
